데이터 저장 구현

TTOzzi opened the pull request:
처음에 Codable 을 채택해서 데이터 저장을 구현했었는데 올려주신 글에서
`상속 관계를 갖고 계층화된 클래스가 다형성으로 동작할 때는 NSCoding 프로토콜을 사용해야 한다.`
라는 걸 보고 NSCoding 을 채택한 방법으로 변경했습니다.
Codable 로 구현했을 때 데이터 저장과 불러오는 것은 잘 되었지만, 데이터를 json 형식으로 인코딩해준 후 그 데이터를 출력해본 결과 최상위 클래스인 Beverage 의 프로퍼티까지만 출력되고 하위 클래스의 프로퍼티들은 출력되지 않는 것을 확인할 수 있었습니다.
NSCoding 을 채택해주기 위해 Struct 로 구현했던 것을 Class 로 바꿔주었는데 저번 피드백에서 Class 와 Struct 를 선택할 때도 다른 개발자를 설득할 수 있어야 한다고 하셨는데 지금처럼 Class 에서만 사용할 수 있는 기능을 쓰기 위해 바꿔준 것도 타당한 이유가 될 수 있을까요?

TTOzzi committed:
VendingMachineManager 객체를 AppDelegate 로 이동

UIApplication 의 shared 클래스를 통해 AppDelegate 의 manager 프로퍼티에 접근

TTOzzi committed:
VendingMachine, Stock 구조체를 클래스로 변경

NSCoding 으로 데이터를 저장해주기위해 클래스로 변경해줌
VendingMachine 의 stock 이 클래스로 바뀌어 stock 내부의 배열에 변화가 생겨도 프로퍼티옵저버가 감지하지 못함
변경사항에 맞게 코드 수정

TTOzzi committed:
VendingMachine 클래스와 그에 포함된 모든 클래스 NSCoding 프로토콜 채택

TTOzzi committed:
데이터 불러오기, 저장 구현 및 불필요 코드 제거

앱이 실행되는 시점에 저장된 데이터를 불러오게 구현
앱이 종료되는 시점에 데이터를 저장하도록 구현

godrm commented:
> NSCoding 을 채택해주기 위해 Struct 로 구현했던 것을 Class 로 바꿔주었는데 저번 피드백에서 Class 와 Struct 를 선택할 때도 다른 개발자를 설득할 수 있어야 한다고 하셨는데 지금처럼 Class 에서만 사용할 수 있는 기능을 쓰기 위해 바꿔준 것도 타당한 이유가 될 수 있을까요?

도저히 안되는 요소가 있으면 선택할 수도 있겠죠. 
클래스여야 해서 해당 프로토콜이 필요한건지, 해당 프로토콜이 필요해서 클래스여야 하는지 헷갈리지 않으면 됩니다.

godrm commented on VendingMachineApp/VendingMachineApp/Model/VendingMachine.swift:
```diff
@@ -35,33 +31,49 @@ struct VendingMachine {
         self.stock.add(beverage: Cola(brand: "코카콜라", amount: 250, price: Money(amount: 1200), name: "코카콜라 제로", calorie: 0, saleablePeriod: 600, sugarContent: 0))
         self.stock.add(beverage: Cantata(brand: "롯데", amount: 500, price: Money(amount: 2500), name: "칸타타 콘트라베이스", calorie: 20, saleablePeriod: 150, caffeineContent: 179, isHot: false))
         self.stock.add(beverage: BananaMilk(brand: "빙그레", amount: 240, price: Money(amount: 1000), name: "바나나맛우유", calorie: 208, saleablePeriod: 90, fatRatio: 15, bananaContent: 3))
         self.stock.add(beverage: Fanta(brand: "코카콜라", amount: 250, price: Money(amount: 1000), name: "환타 오렌지", calorie: 136, saleablePeriod: 600, sugarContent: 34, flavor: "오렌지"))
```
특정한 객체를 채워넣는 것을 자판기 내부에 두기 보다는 별도로 두는 게 어떨까요?
자판기 클래스는 일반적으로 활용할 수 있고, 어떤 음료 객체를 채우는 지는 상황마다 다를 수 있으니까요. 
이 코드가 클래스 동작과 관련없이 바뀔 수 있다는 의미입니다.

godrm commented on VendingMachineApp/VendingMachineApp/Model/Beverage.swift:
```diff
@@ -37,25 +37,51 @@ class Beverage {
+        coder.encode(price, forKey: "price")
+        coder.encode(name, forKey: "name")
+        coder.encode(calorie, forKey: "calorie")
+        coder.encode(isHot, forKey: "isHot")
```
인코딩/디코딩에서 사용하는 키도 하드코딩하지 말고 상수로 표현하는 게 좋습니다.
그래야 같은 키값을 잘못 입력하거나 실수를 줄일 수 있습니다. 참고하세요.

godrm commented on VendingMachineApp/VendingMachineApp/AppDelegate.swift:
```diff
@@ -10,23 +10,15 @@ import UIKit
 
     func application(_ application: UIApplication, didFinishLaunchingWithOptions launchOptions: [UIApplication.LaunchOptionsKey: Any]?) -> Bool {
+        manager.loadData()
         return true
```
이렇게 하면 앱이 백그라운드에 들어갔다 나올 때는 로드가 안되는 경우가 있지 않을까요?

TTOzzi replied:
앱 생명주기에 대해 좀 더 찾아보았는데 앱이 foreground-active 상태에서 벗어날때 데이터를 저장해야하는건 이해가 되는데 앱이 백그라운드에 들어갔다 나올때 왜 로드를 해야하는지 모르겠습니다.
앱이 종료되지 않았는데 로드 할 필요가 있나요..?

godrm replied:
오.. 맞습니다. 그걸 명확하게 이해하고 있으면 됩니다.
그러면 반대로 아래 WillTerminate() 항상 불릴까요? 아닐까요? 
백그라운드에 들어가면 Terminate가 되나요 안되나요? 
이 상태라면 저장이 안되는 경우가 있을까요? 없을까요?

TTOzzi replied:
앱이 백그라운드에 들어간 후 추가적인 작업이 없으면 자동으로 suspended 상태가 됩니다.
그 후 시스템이 메모리가 부족하면 suspended 상태가 된 앱들을 종료시키는데 이때는 willTerminate() 가 안 불린다고 합니다. 그래서 foreground-active 상태에서 벗어날 때 데이터를 저장해주어야 하는 것 같습니다.

godrm commented on VendingMachineApp/VendingMachineApp/Model/Beverage.swift:
```diff
@@ -37,25 +37,51 @@ class Beverage {
+            let price = coder.decodeObject(forKey: "price") as? Money,
+            let name = coder.decodeObject(forKey: "name") as? String,
+            let manufacturingDate = coder.decodeObject(forKey: "manufacturingDate") as? Date,
+            let expirationDate = coder.decodeObject(forKey: "expirationDate") as? Date else { return nil }
```
다른 의도가 있는지 모르겠는데,
디코드가 실패하면 기본값을 가진 객체를 만드는 방법도 있습니다. 실패하는 게 꼭 필요하던가요?

TTOzzi replied:
기본값을 가진 객체를 만들도록 수정해보겠습니다.

TTOzzi replied:
기본값을 가진 객체를 만들어주는 게 디코드 실패 시 디코드를 실패한 프로퍼티에 기본값을 넣어주라는 뜻인가요? 구현하려고 보니 뭔가 이상한 것 같아서요..
애초에 디코드를 실패하는 상황이 있으면 안 될 것 같은 데 그냥 강제언래핑을 해버리는 방법은 어떤가요?

godrm replied:
간혹 있을 수 있습니다. 예를 들어 마지막에 저장한 것과 달리 앱이 업데이트 된 상태로 시작하는데 모델이 달라지는 경우도 있습니다. 저장된 것과 다를수 있다는거죠
그런 경우는 있는 값으로 인스턴스를 표시하는게 좋지 않을까요?
강제 언래핑을 하면 앱이 죽을 수도 있습니다. 안전하게 처리하는게 좋죠
항상 앱이 안 죽고 사용자에게 알려줄 방법이 있으면 그게 더 좋은 선택입니다. 어떤가요?

TTOzzi replied:
음 그렇군요!
알겠습니다. 구현해보겠습니다!

TTOzzi committed:
기본 음료 채우는 코드 위치 변경

VendingMachineManager 가 생성될 때 추가하도록 변경

TTOzzi committed:
데이터 저장 시점 변경

앱이 active 상태에서 벗어날 때 데이터를 저장하도록 수정
iOS 13 이상, 이하 둘 모두 작동하도록 AppDelegate 와 SceneDelegate 둘 다 코드를 작성해줌

TTOzzi committed:
음료 관련 객체 디코드 실패 시 기본값 지정

디코드 실패 시 soldOut 이미지가 출력되도록 해줌

TTOzzi committed:
VendingMachine, Stock, Money 디코드 실패 시 기본값 지정

자판기는 빈 자판기, 재고는 빈 배열, 돈은 0원으로 기본값을 지정해줌

TTOzzi committed:
NSCoder extension 구현

키값을 enum 으로 만들어 그 키값으로 인코딩, 디코딩하도록 구현

TTOzzi commented:
피드백 반영하여 수정 완료했습니다.
간단한 문제로 오랜 시간 헤매는 일이 많았는데 구현하기 전에 문서를 좀 더 꼼꼼히 읽도록 하겠습니다.
같은 키값을 만드는 경우가 생기지 않도록 한눈에 확인하기 위해 키값들을 한 파일에 뒀는데 이걸 키값을 쓰는 곳에 나눠서 작성하는 게 좋을까요?

godrm commented:
> 피드백 반영하여 수정 완료했습니다.
> 간단한 문제로 오랜 시간 헤매는 일이 많았는데 구현하기 전에 문서를 좀 더 꼼꼼히 읽도록 하겠습니다.
> 같은 키값을 만드는 경우가 생기지 않도록 한눈에 확인하기 위해 키값들을 한 파일에 뒀는데 이걸 키값을 쓰는 곳에 나눠서 작성하는 게 좋을까요?

네 클래스에서 각자 사용한 것만 키를 갖고 있는게 더 좋겠죠. 꼭 모아놓는다고 좋은 건 아니에요. 
파일 단위에서도 변화가 최소화하는 게 중요합니다. 참고하세요.

TTOzzi commented:
넵 감사합니다
